### ISSUE TITLE
Broken search in Section when section's VA is less than section's Offset

### ISSUE THREAD

**Issue by ricardoquesada** [radare/radare2#3788]
**UDPATE**: Detailed description of this bug is in this comment: https://github.com/radare/radare2/issues/3788#issuecomment-163042481

Checkout latest radare2-regressions repo and `cd` to:

```
$ cd radare2-regressions/bins/vsf
$ r2 ./c64-rambo2-rom.vsf
```

It has 3 sections, by default the last section is enabled, RAM.

```
[0x0000c2cd]> S
[00] . 0x0001209d mr-x va=0x0000a000 sz=0x2000 vsz=0x2000 BASIC
[01] . 0x0001009d mr-x va=0x0000e000 sz=0x2000 vsz=0x2000 KERNAL
[02] * 0x00000084 mrwx va=0x00000000 sz=0x10000 vsz=0x10000 RAM
```

Now, I will try to search for something that I know is RAM:

```
[0x0000c2cd]> s 0
[0x00000000]> px 32
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000000  ff14 0000 0d0e 000e 0007 0007 0000 0001  ................
0x00000010  0000 0000 0000 0000 0000 0000 0000 0000  ................
```

So, I search for `0xff` and this happens

```
[0x00000000]> /x ff
Searching 1 bytes...
#7 [0xa000-0x10000]
hits: 900
0x0000a336 hit0_0 ff
0x0000a449 hit0_1 ff
0x0000a47f hit0_2 ff
0x0000a491 hit0_3 ff
...
```

**BUG 1**: The range is from 0xa000 to 0x10000, and not from 0x0000 to 0x10000.

Now I will delete the sections BASIC and section KERNAL so that only section RAM is present:

```
[0x0000c2cd]> S
[00] . 0x0001209d mr-x va=0x0000a000 sz=0x2000 vsz=0x2000 BASIC
[01] . 0x0001009d mr-x va=0x0000e000 sz=0x2000 vsz=0x2000 KERNAL
[02] * 0x00000084 mrwx va=0x00000000 sz=0x10000 vsz=0x10000 RAM
[0x0000c2cd]> s 0xa000
[0x0000a000]> S
[00] * 0x0001209d mr-x va=0x0000a000 sz=0x2000 vsz=0x2000 BASIC
[01] . 0x0001009d mr-x va=0x0000e000 sz=0x2000 vsz=0x2000 KERNAL
[02] . 0x00000084 mrwx va=0x00000000 sz=0x10000 vsz=0x10000 RAM
[0x0000a000]> S-
[0x0000a000]> s 0xe000
[0x0000e000]> S
[00] * 0x0001009d mr-x va=0x0000e000 sz=0x2000 vsz=0x2000 KERNAL
[01] . 0x00000084 mrwx va=0x00000000 sz=0x10000 vsz=0x10000 RAM
[0x0000e000]> S-
[0x0000e000]> S
[00] * 0x00000084 mrwx va=0x00000000 sz=0x10000 vsz=0x10000 RAM
```

Now I will search for the first 2 bytes that are in RAM:

```
[0x00000000]> px 2
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000000  ff14
[0x00000000]> /x ff14
Searching 2 bytes...
#7 [0x0-0x10000]
hits: 3
0x000099cd hit0_0 2085
0x0000a691 hit0_1 00ff
0x0000b691 hit0_2 00ff
```

**BUG 2**: I can't find "ff14" at 0x0... and also it reports findings that I'm not interested in like `2085`, `00ff` and `00ff`

However, the range seems to be fixed.

These are my esil search settings:

```
[0x00000000]> e~search
anal.searchstringrefs = false
search.align = 0
search.chunk = 0
search.contiguous = true
search.count = 0
search.distance = 0
search.esilcombo = 8
search.flags = true
search.from = 0xffffffffffffffff
search.in = file
search.kwidx = 1
search.maxhits = 0
search.prefix = hit
search.show = true
search.to = 0xffffffffffffffff
```

**Comment by radare** [comment]
This is probably an io bug. Can u add a test. Can u try with smaller blocksize?

**Comment by ricardoquesada** [comment]
Sure.

More info: Apparently search is searching the physical address and not the virtual address.

I have a section that starts at offset 0x84.
```
[0x00000000]> S
[00] * 0x00000084 mrwx va=0x00000000 sz=0x10000 vsz=0x10000 RAM
```

And the data that is at 0x84 is:
```
[0x00000000]> px 32 @ 0x84
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000084  0000 0000 0000 ac6c 6c68 6700 1200 0000  .......llhg.....
0x00000094  0000 0000 4b0f 0f09 0900 0f00 0000 0000  ....K...........
```

So, if  I search for "ac6c6c68", it should appear at 0x8a:
```
[0x00000000]> /x ac6c6c68
Searching 4 bytes...
# 7 [0x0-0x10000]
hits: 1
0x00000006 hit5_0 000e0007
```

It found the info at 0x6 instead... which is 0x84 (physical offset) + 0x6 (found address) = 0x8c... where it really is.

And what do I have at 0x6...
```
[0x00000000]> px 4 @ 0x6
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00000006  000e 0007                                ....
```

So, it is finding the bytes correctly (at least internally), but it reports the finding in the incorrect place.
A  `finding += physical_address` is missing somewhere.

**Comment by ricardoquesada** [comment]
**More info 2**:
If I try to find something that is between 0 and 0x84  (remember that the physical address is 0x84), it won't find it.

**Comment by ricardoquesada** [comment]
Two regressions tests here:
https://github.com/radare/radare2-regressions/pull/273/files

* Test one tests the physical address bug
* Test two tests the incorrect section bug

**Comment by ricardoquesada** [comment]
I could try to fix this bug myself... but I need a few pointers like, which files/functions to look at. thanks.

**Comment by radare** [comment]
did you tried what i suggested? changing block size and so on? because you can reduce the search for the bug a lot with those hints. i have no time right now for testing this issue

**Comment by ricardoquesada** [comment]
How can I change the block size ? I "grepped" esil, but I couldn't find any "blocksize" option...  I tried searching for `e~block`, `e~size`, `e~bs` but I couldn't find it... However I found `search.chunk`... is that it ?
IIRC I changed the block size before, but for some strange reason I cannot find that option anymore.

Anyway, if you tell me more or less where the bug could be (which file/function), I might be able to fix it.
thanks.

**Comment by radare** [comment]
b 8

**Comment by ricardoquesada** [comment]
thanks. I tried with `b 8`, `b 1` and `b 0x1000`...  I confirm that the block size doesn't fix  / cannot be used as a workaround.

The bugs seems are:

- bug 1: searches on the first section (even if it is not the active one)
- bug 2: when having only one section and also using virtual address, "search" finds the data correctly, but it reports the found data incorrectly. It uses the physical address (and not the virtual address) to report the findings.

I kind of need this functionality, so if none of the radare team can work on this right now, I'm willing to fix it myself... but I need to some help: which files/functions should I take a look at. Thanks.

**Comment by radare** [comment]
Did you tried with different search.in options? maybe search.in=raw can help too. I'll try to find some time tomorrow to give you some pointers on which functions to look at, etc.

**Comment by ricardoquesada** [comment]
@radare thanks.

- bug 1: can be worked-around by using `search.in=io.sections` or `io.section`. Default is `file`, and with `file` the bug can be triggered.
- bug 2: it is still reporting the findings with the incorrect value/offset when `e io.va=true`. If `io.va=false`, then the bug cannot be triggered, but I need to use virtual addresses.

This bug can be triggered with any file + any arch. I just opened a random binary file with r2, added a section manually, enabled `io.va=true` and I triggered the bug.

**Comment by ricardoquesada** [comment]
So the bug was triggered when using virtual address + mapped sections.
I added a possible bug fix. Not sure if that is the best way to fix it though.

**Comment by radare** [comment]
Can you add a test case too? Thanks!

**Comment by ricardoquesada** [comment]
yes, I already added the test case... it is the first test of this file:
https://github.com/radare/radare2-regressions/blob/master/t/search/bug_3788#L4

With this patch, that test passes.

**Comment by radare** [comment]
Thanks. Im afk will hopefully review tonight, just replying from phone :)

**Comment by ricardoquesada** [comment]
ok. I've just sent another PR that removes the "broken" from that test case:
https://github.com/radare/radare2-regressions/pull/279

**Comment by radare** [comment]
Fixed by @ricardoquesada Thanks!

**Comment by ricardoquesada** [comment]
cmd.hit, searchflags had the same issue. Sent new PR: https://github.com/radare/radare2/pull/3813

**Comment by radare** [comment]
yeah merged both :)

**Comment by ricardoquesada** [comment]
I found more info:

- The pull requests that I sent break other stuff, specially nes binary support. I'll add a test case for it.
- I found more info about this bug:

The bug can be triggered when:

- Using a mapped file
- The Virtual Address collides with the loaded file.

As an example, when I load a file, the file will be loaded at address 0. But if also use 0x0 as the virtual address, this bug will be triggered.

**Comment by ricardoquesada** [comment]
Test case that test when VA doesn't collide with its header is here: https://github.com/radare/radare2-regressions/pull/282

**Comment by radare** [comment]
I don't see why revert a fix, even if the fix is not correct for all cases, the solution should be to push the correct patch and not mark the tests as broken

**Comment by ricardoquesada** [comment]
your call. Basically my patch breaks existing code, and apparently nobody triggered the bug that I was trying to fix. That's why I thought it was good to revert it.
Of course sending the correct patch is what I would like to do... I'm working on that, but I'm not sure if I'll be able to fix it. It seems that the bug is in the io_seek logic (but I could be wrong), and I don't totally understand the internals of the io code.

**Comment by ricardoquesada** [comment]
**TLDR;**
`r_io_section_vaddr_to_maddr_try(io, vaddr)` fails if vaddr is both a valid file address and VA.
That happens when `section->va` < `section->offset`

**Description of the bug**:

I tried to fix the bug, but I couldn't fix it without breaking other parts of radare2. A more knowledgeable guy is needed to fix this bug.

Basically `r_io_read_at` is using file addresses to read the memory.
Internally `r_io_read_at` calls `r_io_seek` which accepts file address and VAs. `r_io_seek` will try to guess if the address that is being passed is memory address or a VA.
Internally `r_io_seek` calls `r_io_section_vaddr_to_maddr_try` in other to know whether or not the address is a VA or a file address.
The problem is that that file fails to determine if an address like 80 is a file address or VA.

One possible solution is to add a new API, like `r_io_map_select(io, addr, addr_type)`, where `addr_type` says whether it is a File Address or a VA.

And perhaps `r_io_map_select` should call interally `r_io_map_select()` but I'm not sure about this.

**Comment by radare** [comment]
@condret

**Comment by ricardoquesada** [comment]
@radare @condret  I've just updated the description of the bug. I added more findings and a possible patch. Just ignore my comments about `r_io_section_vaddr_to_maddr_try`.

**Comment by radare** [comment]
this was merged. so i guess this issue is gone too
